### ISSUE TITLE
Broken (read-only) database connections not getting removed from connection pool, keycloak claims to be healthy.

### ISSUE THREAD

**Issue by schrom** [keycloak/keycloak#24493]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

We have a master/standby setup of our postgres database (standard zalando/postgres-operator).

We had a "hiccup" in our k8s cluster, which resulted in  a failover - master and standby got switched. The peculiar thing about this switch was that the former master survived the hiccup, only the former standby got restarted and promoted to master.

This resulted in all connections to the database in the Keycloak pool still being active, but now pointing to the standby database, meaning they are read-only.

From that moment on, Keycloak stopped working properly. Every login failed with this exception:

```
2023-10-31 18:47:47.096	
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
2023-10-31 18:47:47.096	
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
2023-10-31 18:47:47.096	
	at com.sun.proxy.$Proxy86.executeUpdate(Unknown Source)
2023-10-31 18:47:47.096	
	at org.postgresql.ds.PGPooledConnection$StatementHandler.invoke(PGPooledConnection.java:441)
2023-10-31 18:47:47.096	
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-10-31 18:47:47.096	
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-10-31 18:47:47.096	
	at jdk.internal.reflect.GeneratedMethodAccessor181.invoke(Unknown Source)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:152)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496)
2023-10-31 18:47:47.096	
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
2023-10-31 18:47:47.096	
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
2023-10-31 18:47:47.096	
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)

2023-10-31 18:47:47.096	
Caused by: org.postgresql.util.PSQLException: ERROR: cannot execute INSERT in a read-only transaction
```

This could only be resolved by manually restarting the keycloak server to force a reconnection to the new master database.

### Version

quay.io/keycloak/keycloak:20.0.5 installed via helm chart

### Expected behavior

Database connection that cause exceptions should be removed from the pool and not be reused. No manual intervention should be necessary.

### Actual behavior

Keycloak is using the non-functional database connections ad inifinitum and doesn't care about exceptions. Health check is reporting "healthy", though the database connection is read only.

### How to Reproduce?

If you do not want to setup the whole HA feature, start Keycloak with a local Postgres and set it to read only: 
`alter system set default_transaction_read_only=on;`

### Anything else?

We have a lot of Spring application in our cluster. They are using the Hikari connection pool. This pool can be configured so that database connections which run into an exception are always and immediately removed from the connection pool. This eventually auto-heals the application, as the read only connection get dropped and replaced by fresh connections to the master database.

As Keycloak is using the Agroal pool, I found this is the documenation:

> exceptionSorter(ExceptionSorter) - This extension point tells Agroal which exceptions indicate that the connection is in an invalid state and should be removed from the pool. Agroal provides a few implementations of the ExceptionSorter interface for selected databases. By default, no exception is considered fatal.

It would be nice if you could activate this feature.

**Comment by shawkins** [comment]
> It would be nice if you could activate this feature.

@schrom @luringens Double checking with the existing PostgreSQLExceptionSorter, it is looking specifically for 08xxx SQLStates.  In this case the SQLState should be 25006 or similar.  I don't see a documented way through quarkus of adding a custom sorter.  Perhaps a workaround would be to add validation sql that will error when read-only "SELECT case when pg_is_in_recovery() then 1/0 end"

**Comment by uday-accel** [comment]
Any update one this? we are also seeing same issue in our cluster.

**Comment by ahus1** [comment]
I understand the note that the exception sorter "PostgreSQLExceptionSorter" is part of Agroal. If it needs to change (which I assume), we would need to wait for another release of that one plus of Quarkus? 

Still, I don't like the situation where I close a connection once an error occurs. Ideally this check if it is readonly should be done when a connection is took from the pool (start of the lease), so we have to discard it and possibly create a new one before starting any work on it. This avoids showing an exception to a user. 

I see that the interface `java.sql.Connection` has a method `isReadOnly()`. Looking at the code of the different drivers, this is a local attribute on a connection instance and nothing that reads the current status from the database. As an alternative, I see that other DB pools like [Hikari](https://github.com/brettwooldridge/HikariCP) have a `connectionTestQuery`.  Haven't seen support for such a query in Agroal, don't know if I missed something there. It would come with the downside of an additional SQL query at the beginning of each JDBC connection lease. 

For AWS, there is a JDBC Wrapper which handles this in a more proactive way, see #26496 for the progress. Are similar wrappers available for other clustered databases as well?

Having said all that, closing connections which we know are broken could still be the first step, but possibly not the final state we're looking for.

**Comment by mhajas** [comment]
I am removing the triage tag as this is indeed an issue. However, the solution here is not straightforward. 

We tested Keycloak with Amazon Aurora PostgreSQL with two instances writer and reader instance. During failover (setting writer->reader, reader->writer) we encountered the issue with read-only transactions. The issue is that already acquired connections are still pointing to the previous writer (now reader) instance which is now read-only. 

Now there are more ways to solve it:
1. AWS provides a [JDBC Wrapper ](https://github.com/awslabs/aws-advanced-jdbc-wrapper) that solves this issue and Keycloak can recover from the error state. There may be more wrappers like this for other databases.
2. As suggested by @shawkins, it is possible to use [quarkus.datasource.jdbc.validation-query-sql](https://quarkus.io/guides/datasource#quarkus-agroal_quarkus.datasource.jdbc.validation-query-sql) and I can confirm this works and Keycloak can recover from the error state. However, the SQL query `SELECT case when pg_is_in_recovery() then 1/0 end` did not work for me. I was always getting an exception even with: `SELECT case when pg_is_in_recovery() then 1 else 1/0 end`. I confirmed this works with the query `SELECT ID FROM public.databasechangeloglock WHERE ID=1000 FOR UPDATE` but this is **NOT** recommended for usage in production as we never properly tested all the consequences this config may have. We used it just to confirm this config option can solve the issue when a good query is chosen.
3. Another option is to not check the connections are ok but only check this after a failure happens. As already mentioned this can be done by an `ExceptionSorter`, however, the current sorter within Agroal for PostgreSQL only fails for `08xxx` error and the read-only error is `25xxx` which means the connection is not dropped when it is encountered. This would require a change in Agroal or some mechanism that allows adding more sorters or similar. The disadvantage of this approach is there still will be `[number-of-connections]` failures that users see before Keycloak recovers from this state.

Probably the best solution is number 1. but it depends on the database used as not all may support this.

**Comment by ahus1** [comment]
@mhajas - thank you for the investigation. It seems that PostgreSQL (or at least Aurora?) does an eager evaluation of the '1/0' if it is used as a constant. 

I tried to rewrite it to prevent that eager evaluation, and the following statement worked for me on an Aurora Multi-AZ database. It succeeded on the writer note, and failed on the reader node:

```
SELECT 1 / case when pg_is_in_recovery() then 0 else 1 end
```

As this statement doesn't run a `... FOR UPDATE`, it shouldn't lock any tables or rows, which should perform better. It would be good if the community (@schrom / @luringens) could comment if this works for you.

About the "help wanted" tag: It would be great if the community would create an issue against Agroal, start a discussion that `25006` should close connections, and then possibly create a pull request. If you have an issue, please link it here so people can follow the link, upvote it and comment.

Project link: https://agroal.github.io/

**Comment by aleksey-masl** [comment]
Colleagues, we have the same question. Any ideas?

**Comment by aleksey-masl** [comment]
It work for me in yandex-cloud

db-url=jdbc:postgresql://rc1a-***.mdb.yandexcloud.net:6432,rc1b-***.mdb.yandexcloud.net:6432,rc1d-***.mdb.yandexcloud.net:6432/keycloak?targetServerType=master

**Comment by ahus1** [comment]
Created an upstream issue to get input from the Agroal project: https://issues.redhat.com/browse/AG-236

**Comment by ahus1** [comment]
After some more feedback I received, I understand that `targetServerType` is a standard option of the PostgreSQL driver: https://jdbc.postgresql.org/documentation/use/

IMHO it makes sense for Keycloak to add this to the JDBC URL automatically if the option `targetServerType` has not been set. See 

https://github.com/keycloak/keycloak/blob/99478887a47796ecd65ae7347c9ffd44fe7f6755/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L181-L183

on how we do something similar for H2 (note that this is called in two places in the code). 

I'd be happy to see a pull request for this. Note that the latest version of the PostgreSQL driver now prefers "primary" instead of "master".

**Comment by ahus1** [comment]
After some more feedback I received, I understand that `targetServerType` is a standard option of the PostgreSQL driver: https://jdbc.postgresql.org/documentation/use/

IMHO it makes sense for Keycloak to add this to the JDBC URL automatically if the option `targetServerType` has not been set. See 

https://github.com/keycloak/keycloak/blob/99478887a47796ecd65ae7347c9ffd44fe7f6755/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L181-L183

on how we do something similar for H2 (note that this is called in two places in the code). 

I'd be happy to see a pull request for this. Note that the latest version of the PostgreSQL driver now prefers "primary" instead of "master".

Side note: The check of a read-only database happens in this case only when a connection is established, not after the connection has been created. It would need testing with a real switchover to see if this is sufficient for real-world switchovers, and it would assume that a database would close all existing connections during a switchover.

**Comment by ahus1** [comment]
gpcmol - as you mentioned in another issue that you're using a PostgreSQL cluster, I wonder if you have been using the `targetServerType=primary` in your JDBC setup and if this would be a good default for Keycloak. It could IMHO resolve a situation where DNS entries are lagging behind, and Keycloak would connect to a read-only node accidentially. I would hope that with this `targetServerType`, KC would then drop the connection again.

**Comment by ahus1** [comment]
Thank you for checking. 

When testing with Aurora on AWS we've seen problems when just switching the writer instance to another instance. As then the current writer instance it turned read-only, AFAIK Keycloak would try to connect to it while the DNS hasn't switched. 

So while it doesn't affect you when killing a Pod, so this would just be an additional safeguard then.

**Comment by eicki** [comment]
Just as a side note for all who use the aws-advanced-jdbc-wrapper to connect to Aurora. It includes a failover plugin that works perfectly and is enabled by default ... unless you set the wrapperPlugins for other purposes, e.g. iam authentication. So in this case you need to set it to wrapperPlugins=iam,failover. Otherwise you will get the same issues as described here.

https://github.com/aws/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/using-plugins/UsingTheFailoverPlugin.md#using-the-failover-plugin

**Comment by ahus1** [comment]
@eicki - can you please provide an example of the before/after configuration that first didn't work and then worked. Maybe I can add something to Keycloak that we spit out a warning if this  configuration occurs.

**Comment by ahus1** [comment]
I figured out that the original idea to add the property by default doesn't work as expected. I created a follow-up issue #40383 to rectify this.
